interfaces/builtin/utils.go

```diff
@@ -125,3 +126,49 @@ func verifySlotPathAttribute(slotRef *interfaces.SlotRef, attrs interfaces.Attre
 	}
 	return cleanPath, nil
 }
+
+// aareExclusivePatterns takes a string and generates deny alternations. Eg,
+// aareExclusivePatterns("foo") returns:
+// []string{
+//   "[^f]*",
+//   "f[^o]*",
+//   "fo[^o]*",
+// }
+func aareExclusivePatterns(orig string) []string {
+	s := make([]string, len(orig))
+	prefix := ""
+	for i, letter := range orig {
+		if i > 0 {
+			prefix = orig[:i]
+		}
+		s[i] = fmt.Sprintf("%s[^%c]*", prefix, letter)
+	}
+	return s
+}
+
+// getDesktopFileRules(<snap instance name>) generates snippet rules for
+// allowing access to the specified snap's desktop files in
+// dirs.SnapDesktopFilesDir, but explicitly denies access to all other snaps'
+// desktop files since xdg libraries may try to read all the desktop files
+// in the dir, causing excessive noise. (LP: #1868051)
+func getDesktopFileRules(snapInstanceName string) []string {
+	baseDir := dirs.SnapDesktopFilesDir
+
+	rules := []string{
+		"# Support applications which use the unity messaging menu, xdg-mime, etc",
+		"# This leaks the names of snaps with desktop files",
+		fmt.Sprintf("%s/ r,", baseDir),
+		"# Allowing reading only our desktop files (required by (at least) the unity",
+		"# messaging menu).",
+		"# parallel-installs: this leaks read access to desktop files owned by keyed",
+		"# instances of @{SNAP_NAME} to @{SNAP_NAME} snap",
+		fmt.Sprintf("%s/@{SNAP_INSTANCE_NAME}_*.desktop r,", baseDir),
+		"# Explicitly deny access to other snap's desktop files",
+		fmt.Sprintf("deny %s/@{SNAP_INSTANCE_NAME}[^_.]*.desktop r,", baseDir),
+	}
+	for _, t := range aareExclusivePatterns(snapInstanceName) {
```

Comment:
Looking at the code above, `snapInstanceName` comes from `snap.Info.InstanceName()`, which is of the form `$snapName_$instanceKey`.

However, desktop files are named with the prefix `$snapName+$instanceKey` (see `desktopPrefix` in `wrappers/desktop.go`).  Surely this is the string we want to exclude?

Thinking more about this, the same concern probably applies to rules granting access to desktop files based on `@{SNAP_INSTANCE_NAME}`.

Maybe the `desktopPrefix` helper in wrappers should be hoisted to a method on `snap.Info`, and used here and maybe `interfaces/apparmor/template_vars.go`.

Reply:
@jhenstridge - good catch. I'll adjust and then take a look at the policy.

Reply:
I've created https://github.com/snapcore/snapd/pull/9167 to address this issue. I'll adjust this PR to pull in those commits and use the new info.DesktopPrefix() method.